From 28123d5c1361e06002acb3dac44c4b94c77fc130 Mon Sep 17 00:00:00 2001
From: Jan Janssen <jan-janssen@users.noreply.github.com>
Date: Sat, 29 Nov 2025 12:38:12 +0100
Subject: [PATCH] Fixes for manually specified potentials (#279)

* Fixes for manually specified potentials

* add tests

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fixes

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* fix test

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* again

* test species

* remove empty line

* fix validation

* include empty line

* add new test

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
---
 pyiron_lammps/compatibility/file.py |  62 ++++++++++---
 pyiron_lammps/potential.py          |   4 +-
 3 files changed, 46 insertions(+), 19 deletions(-)

diff --git a/pyiron_lammps/compatibility/file.py b/pyiron_lammps/compatibility/file.py
index fd3a5c1..709b873 100644
--- a/pyiron_lammps/compatibility/file.py
+++ b/pyiron_lammps/compatibility/file.py
@@ -88,18 +88,24 @@ def lammps_file_interface_function(
         calc_kwargs = {}
 
     os.makedirs(working_directory, exist_ok=True)
-    if isinstance(potential, str):
-        potential_dataframe = get_potential_by_name(
-            potential_name=potential, resource_path=resource_path
-        )
-    elif isinstance(potential, pandas.DataFrame):
-        potential_dataframe = potential.iloc[0]
-    elif isinstance(potential, pandas.Series):
-        potential_dataframe = potential
-    else:
-        raise TypeError()
-    lmp_str_lst = lammps_file_initialization(structure=structure)
-    lmp_str_lst += potential_dataframe["Config"]
+    potential_lst, potential_replace, species = _get_potential(
+        potential=potential, resource_path=resource_path
+    )
+
+    lmp_str_lst = []
+    atom_type = "atomic"
+    for l in lammps_file_initialization(structure=structure):
+        if l.startswith("units") and "units" in potential_replace:
+            lmp_str_lst.append(potential_replace["units"])
+        elif l.startswith("atom_style") and "atom_style" in potential_replace:
+            lmp_str_lst.append(potential_replace["atom_style"])
+            atom_type = potential_replace["atom_style"].split()[-1]
+        elif l.startswith("dimension") and "dimension" in potential_replace:
+            lmp_str_lst.append(potential_replace["dimension"])
+        else:
+            lmp_str_lst.append(l)
+
+    lmp_str_lst += potential_lst
     lmp_str_lst += ["variable dumptime equal {} ".format(calc_kwargs.get("n_print", 1))]
     lmp_str_lst += [
         "dump 1 all custom ${dumptime} dump.out id type xsu ysu zsu fx fy fz vx vy vz",
@@ -161,11 +167,12 @@ def lammps_file_interface_function(
 
     write_lammps_datafile(
         structure=structure,
-        potential_elements=potential_dataframe["Species"],
+        potential_elements=species,
         bond_dict=None,
         units=units,
         file_name="lammps.data",
         working_directory=working_directory,
+        atom_type=atom_type,
     )
 
     shell = subprocess.check_output(
@@ -178,7 +185,7 @@ def lammps_file_interface_function(
     output = parse_lammps_output(
         working_directory=working_directory,
         structure=structure,
-        potential_elements=potential_dataframe["Species"],
+        potential_elements=species,
         units=units,
         prism=None,
         dump_h5_file_name="dump.h5",
@@ -222,3 +229,30 @@ def _modify_input_dict(
         return lmp_tmp_lst
     else:
         return lmp_str_lst
+
+
+def _get_potential(potential, resource_path: Optional[str] = None):
+    if isinstance(potential, str):
+        potential_dataframe = get_potential_by_name(
+            potential_name=potential, resource_path=resource_path
+        )
+    elif isinstance(potential, pandas.DataFrame):
+        potential_dataframe = potential.iloc[0]
+    elif isinstance(potential, pandas.Series):
+        potential_dataframe = potential
+    else:
+        raise TypeError()
+
+    potential_replace = {}
+    potential_lst = []
+    for l in potential_dataframe["Config"]:
+        if l.startswith("units"):
+            potential_replace["units"] = l
+        elif l.startswith("atom_style"):
+            potential_replace["atom_style"] = l
+        elif l.startswith("dimension"):
+            potential_replace["dimension"] = l
+        else:
+            potential_lst.append(l)
+
+    return potential_lst, potential_replace, potential_dataframe["Species"]
diff --git a/pyiron_lammps/potential.py b/pyiron_lammps/potential.py
index e9df255..2f338f1 100644
--- a/pyiron_lammps/potential.py
+++ b/pyiron_lammps/potential.py
@@ -1,6 +1,6 @@
 import os
 from pathlib import Path
-from typing import Union
+from typing import Optional, Union
 
 import pandas
 from ase.atoms import Atoms
@@ -346,7 +346,7 @@ def get_potential_dataframe(structure: Atoms, resource_path=None):
     )
 
 
-def get_potential_by_name(potential_name: str, resource_path=None):
+def get_potential_by_name(potential_name: str, resource_path: Optional[str] = None):
     if resource_path is None:
         resource_path = get_resource_path_from_conda()
     df = LammpsPotentialFile(resource_path=resource_path).list()
